src/pybind/mgr/mgr_module.py

```diff
@@ -1,10 +1,16 @@
 import ceph_module  # noqa
 
-try:
-    from typing import Set, Tuple, Iterator, Any, Dict, Optional, Callable, List
-except ImportError:
-    # just for type checking
-    pass
+from typing import Set, Tuple, Iterator, Any, Dict, Optional, Callable, List, Union, TYPE_CHECKING
+if TYPE_CHECKING:
+    import sys
+    if sys.version_info >= (3, 8):
+        from typing import Literal
```

Comment:
instead of hardwiring to a certain version of python, i'd suggest just catch `ImportError`.

Reply:
[3.8 was the exact version where `Literal` was added](https://docs.python.org/3/library/typing.html#typing.Literal)... And the try/catch `ImportError` in this case is less informative and would require adding a misleading comment:
```py
try:
    from typing import Literal
except ImportError:
    # Ignore on Python < 3.8
    pass
```

Reply:
> 3.8 was the exact version where Literal was added...

yeah. but it's still a documented magic number.

i am confused, shall we instead have
```py
try:
    from typing import Literal
except ImportError:
    from typing_extensions import Literal
```
?

Reply:
my 2 cents:

* the current solution using `if sys.version_info >= (3, 8):` works for me, as this is part of the standard lib and not subject to change.
* Using 
    ```python
   except ImportError:
        # Ignore on Python < 3.8
        pass
   ```
    works for me as well, as the comment is descriptive
* and omitting the comment would work for me, too. 

IMO this is not something we need to discuss in detail as all alternative are actually good.

Reply:
Thank you for your feedback, @tchaikov and @sebastian-philipp !

This is a complex matter. Not the import block, but how to deal with the actual imported contents (e.g.: `Literal`). `typing-extensions` is aimed to be a backport library for Python >3.6 `typing` additions to 3.5-3.5.  We mostly want this just for the testing phase (where can simply ran everything in a virtualenv with pip-resolved dependencies), but Python <3.9 interprets annotations, so those need to be valid runtime code too. `typing-extensions` packages are available for most of our supported distros, except Ubuntu, where only in 20.04 this is available. Additionally this leads to conditional `Requires` and `BuildRequires` in our code, based on a macro `python3_version_nodots` which is not correctly defined (it should be `36` but defaults to `3`) and will be broken in Python 3.10. And `debian/control` does not even allow for conditional deps...

So I'll take the following approach:
* Assume Python 3.6 `typing`
* Allow >3.6 `typing` types conditionally defined in the `if TYPE_CHECK` block and as quoted annotation in the production code.
* Mypy + >3.6 `typing` will be run in a tox environment (no package dependencies installed).

Reply:
And there was a good reason for the if-else import. `mypy` doesn't seem to work with the try-except import mechanism and recommends the [if-else instead](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks).
```
mgr_module.py:7: error: Module 'typing' has no attribute 'Literal'
```